### PR TITLE
JIT: Assume allocations succeed

### PR DESCRIPTION
**Pull request by jakobbotsch**
We call a NOMEM function when allocations fail that is already marked as no-return, but at least MSVC is not able to pick up on this fact. The result is that all operator new calls end up with an unnecessary null check. This adds an __assume to allow MSVC to recognize that allocations never fail. For other compilers, it uses the preexisting UNREACHABLE() macro, but MSVC was not able to recognize this either.

There was also a workaround for an old VSW bug that I've removed.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=192997&view=ms.vss-build-web.run-extensions-tab)

cc @dotnet/jit-contrib

File: src/coreclr/jit/alloc.h [hunk 1]
```diff
@@ -235,6 +235,16 @@ class CompAllocator
...
 
+#if defined(_MSC_VER)
+        // MSVC does not handle the other pattern correctly to recognize this fact.
+        __assume(p != nullptr);
+#elif !defined(DEBUG)
+        if (p == nullptr)
+        {
+            UNREACHABLE();
+        }
+#endif
```
Comment (tannergooding) [inline src/coreclr/jit/alloc.h:246]: Should there be a general `ASSUME(expr)` helper that abstracts this?

**Review by tannergooding — APPROVED** [review]

**Review by EgorBo — APPROVED** [review]

**Review by BruceForstall — APPROVED** [review]
The TP diffs are amazingly good.

File: src/coreclr/jit/alloc.h (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/alloc.h]: It turns out there is already a `COMPILER_ASSUME` next to `UNREACHABLE` that ends up being `__assume` with the specified expression on MSVC, so I've switched to that.

**Pull request merged**